### PR TITLE
Georeference filter

### PR DESCRIPTION
**Pull request by gui2dev**
# filters.georeference

Given a trajectory (typically sbet file) and a scan2imu transformation, do georeferencing of the data, using `GpsTime Dimension` as a synchronisation reference.
Trajectory must :
  * contains `X`, `Y`, `Z`, `Roll`, `Pitch`, `Yaw`, `WanderAngle` and `GpsTime` ;
  * have coordinates expressed in `WGS84` system (EPSG:4979) ;
  * have all its angle values expressed in radians.

> **_Streamable stage_**
> 
>  This stage supports streaming operations

> **_Dependencies_**
> 
>  ~~This stage requires GeographicLib~~

```JSON
[
  "input.rxp",
  {
    "type" : "filters.georeference",
    "trajectory_file" : "sbet.out",
    "trajectory_options" : { "type" : "readers.sbet", "angle_as_degrees" : false },
    "scan2imu" : "0.547418069126518 0.629350569360448 0.055655 0.282019849017109 0.832472349301833 -0.476921998243165 -0.830192 -0.784992666412172 -0.085575961821537 -0.613566026143419 -0.099056 0.000 0.000 0.000 1.000"
  },
  {
    "type" : "filters.reprojection",
    "in_srs" : "EPSG:4979",
    "out_srs" : "EPSG:2154+5720"
  }
  "output.laz"
]
```

## Options

* `trajectory_file` : path to a sbet trajectory file.
* `trajectory_options` : JSON object containing the trajectory reader options.
* `scan2imu` : 4x4 transformation matrix from scanner frame to body imu.
* `reverse` : revert georeferencing (go back to scanner frame) 
* `time_offset` : timestamp offset between trajectory and scanner GpsTime
*  `coordinate_system` : Two right-handed variants exist for Local tangent plane coordinates: east, north, up (ENU) coordinates and north, east, down (NED). The `scan2imu` matrix can be expressed with Z point up or down.

### CODE REVIEW

**Comment by hobu** [comment]
> This stage requires GeographicLib

Any chance to require just PROJ here? PROJ already has GeographicLib embedded in it. 

The reason I ask is adding GeographicLib as a dependency means that this filter will be a dynamically loaded plugin. If you can base it on PROJ, our hard GDAL requirement (which also has PROJ as a req) will mean it will always be available.

**Comment by gui2dev** [comment]
> > This stage requires GeographicLib
> 
> Any chance to require just PROJ here? PROJ already has GeographicLib embedded in it. https://proj.org/en/latest/geodesic.html

Doable, we are using a `GeographicLib::LocalCartesian` class. i guess there is something similar in `proj`.

File: plugins/CMakeLists.txt [hunk 1]
```diff
@@ -74,3 +74,7 @@ endif()
 if(BUILD_PLUGIN_E57)
     add_subdirectory(e57)
 endif()
+
+if (BUILD_PLUGIN_GEOREFERENCE)
```
Comment (hobu) [inline plugins/CMakeLists.txt:78]: This should be removed now that we're building by default in `./filters`.

**Review by hobu — CHANGES_REQUESTED** [review]

File: filters/GeoreferenceFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+{
+    m_config.reset(new Config(m_trajectory, *m_matrix));
+}
+
+void GeoreferenceFilter::addDimensions(PointLayoutPtr layout)
+{
+    layout->registerDim(DimId::ScanAngleRank);
+}
+
+void GeoreferenceFilter::prepared(PointTableRef table) {}
```
[the remark this answers is not in the document]
Comment (gui2dev) [reply filters/GeoreferenceFilter.cpp:107]: I'm removing that, it's very `rxp` related

**Comment by gui2dev** [comment]
I just tested the `proj` (actually `ogr`) implementation against the `GeographicLib`. It takes quite some time to always reset the `OGRCoordinateTransformation` for each trajectory position.
But I can confirm it's working.
There might be a better approch to direct use a combination of `osgeo::proj::operation` to limit the `gdal` overhead. But it means digging deeper in the `proj` documentation.

Other point, we expect `scan2imu` to be expressed with Z pointing down. Maybe an option ? Or just state that in documentation. What do you think @hobu ?

**Comment by hobu** [comment]
> Other point, we expect scan2imu to be expressed with Z pointing down. Maybe an option ? Or just state that in documentation. What do you think @hobu ?

Maybe provide an option to flip the data for them, unless you think the data is always coming in Z-down.

On the performance thing it would indeed be better to use PROJ API and compose the transformation there instead of going through GDAL. We don't FindProject(PROJ), however. This is assumed as part of having GDAL, so I'm supportive of adding it if you want to take this a step further. That said, where it's at now is useful at least.

**Comment by gui2dev** [comment]
> Maybe provide an option to flip the data for them, unless you think the data is always coming in Z-down.
I would say it's usual to have calibration matrix with Z-down, but adding an option won't hurt.

> On the performance thing it would indeed be better to use PROJ API and compose the transformation there instead of going through GDAL. We don't FindProject(PROJ), however. This is assumed as part of having GDAL, so I'm supportive of adding it if you want to take this a step further. That said, where it's at now is useful at least. 
Ok. I'll just have to figure out how `proj` works.

**Review by hobu — APPROVED** [review]

**Comment by gui2dev** [comment]
@hobu there is a full `proj` implementation.
I'm not sure if I can reach the right operation using the c++ proj api. I still need to create stuff from user input.

N.B. : this implementation requires at least proj 9.3. Because :
> the essential osgeo::proj::operation::CoordinateOperation::coordinateTransformer() method was only added in PROJ 9.3

**Comment by hobu** [comment]
> N.B. : this implementation requires at least proj 9.3. Because :

Frustrating. 

Up to you how to proceed. I'll merge it when you say it is ready.

**Comment by hobu** [comment]
@rouault do you see a way to do this in older PROJ without `osgeo::proj::operation::CoordinateOperation::coordinateTransformer() `?

**Comment by gui2dev** [comment]
> > N.B. : this implementation requires at least proj 9.3. Because :
> 
> Frustrating. 
> 
> Up to you how to proceed. I'll merge it when you say it is ready.

Well the doc says that you have to use C api. So that might be solved with some preprocessing

**Comment by rouault** [comment]
>  do you see a way to do this in older PROJ without coordinateTransformer

use the C API : proj_create() + proj_trans()

**Comment by gui2dev** [comment]
> @rouault do you see a way to do this in older PROJ without `osgeo::proj::operation::CoordinateOperation::coordinateTransformer() `? 

What would be even better, would be to get to the right `CoordinateOperation` without have to use`CreateFromUserInput`.

**Comment by gui2dev** [comment]
@hobu there is a full `C` implementation. No `#if` statement to check if `PROJ_VERSION` >= 9.3.0
Should work with all `PROJ` version I guess.

**Comment by gui2dev** [comment]
There, that's my final commit on that PR. Any file type can be used as trajectory file, plus you can specify options for the reader.

**Comment by xiaodong2077** [comment]
> @hobu there is a full `C` implementation. No `#if` statement to check if `PROJ_VERSION` >= 9.3.0 Should work with all `PROJ` version I guess.

@hobu  @gui2dev
Hey guys, [these statements](https://github.com/PDAL/PDAL/blob/65acaf14e59cdb5d14f765ad64bc61f35d0496c4/filters/private/georeference/LocalCartesian.cpp#L39) are still existed in current branch.
```c+++
#if PROJ_AT_LEAST_VERSION(9, 3, 0)
#include "proj/util.hpp" // for nn_dynamic_pointer_cast
#else
#endif
```

I meet errors when I try to compile it with proj(version 6+).
```
error: missing binary operator before token "("
   39 | #if PROJ_AT_LEAST_VERSION(9, 3, 0)
```

I find [it](https://proj.org/en/9.3/development/reference/macros.html) from proj doc
```
PROJ_AT_LEAST_VERSION(maj, min, patch)
New in version 8.0.1.

Macro that returns true if the current PROJ version is at least the version specified by (maj,min,patch)

Equivalent to PROJ_VERSION_NUMBER >= PROJ_COMPUTE_VERSION(maj,min,patch)
```

So errors came from `PROJ_AT_LEAST_VERSION`.

**Pull request merged**